src/modules/schnorrsig/main_impl.h

```diff
@@ -0,0 +1,312 @@
+            return 0;
+        }
+    }
+    return 1;
+}
+
+/* Helper function for batch verification.
+ * Hashes signature verification data into the randomization seed and initializes ecmult_context. */
+int secp256k1_schnorrsig_verify_batch_init_randomizer(const secp256k1_context *ctx, secp256k1_schnorrsig_verify_ecmult_context *ecmult_context, secp256k1_sha256 *sha, const secp256k1_schnorrsig *const *sig, const unsigned char *const *msg32, const secp256k1_pubkey *const *pk, size_t n_sigs) {
+    size_t i;
+
+    if (n_sigs > 0) {
+        ARG_CHECK(sig != NULL);
+        ARG_CHECK(msg32 != NULL);
+        ARG_CHECK(pk != NULL);
+    }
+
+    for (i = 0; i < n_sigs; i++) {
+        unsigned char buf[33];
+        size_t buflen = sizeof(buf);
+        secp256k1_sha256_write(sha, sig[i]->data, 64);
+        secp256k1_sha256_write(sha, msg32[i], 32);
+        secp256k1_ec_pubkey_serialize(ctx, buf, &buflen, pk[i], SECP256K1_EC_COMPRESSED);
+        secp256k1_sha256_write(sha, buf, 32);
+    }
+    ecmult_context->ctx = ctx;
+    ecmult_context->sig = sig;
+    ecmult_context->msg32 = msg32;
+    ecmult_context->pk = pk;
+    ecmult_context->n_sigs = n_sigs;
+
+    return 1;
+}
+
+/* Helper function for batch verification.
+ * Sums the s part of all signatures multiplied by their randomizer */
+int secp256k1_schnorrsig_verify_batch_sum_s(secp256k1_schnorrsig_verify_ecmult_context *ecmult_context, secp256k1_scalar *s, const secp256k1_schnorrsig *const *sig, size_t n_sigs) {
+    size_t i;
+
+    for (i = 0; i < n_sigs; i++) {
+        int overflow;
+        secp256k1_scalar term;
+        if (i % 2 == 0) {
+            secp256k1_schnorrsig_batch_randomizer(ecmult_context->randomizer_cache, ecmult_context->chacha_seed, i / 2);
+        }
+
+        secp256k1_scalar_set_b32(&term, &sig[i]->data[32], &overflow);
+        if (overflow) {
+            return 0;
+        }
+        secp256k1_scalar_mul(&term, &term, &ecmult_context->randomizer_cache[i % 2]);
+        secp256k1_scalar_add(s, s, &term);
+    }
+    return 1;
+}
+
+/* schnorrsig batch verification.
+ * Seeds a random number generator with the inputs and derives a random number ai for every
+ * signature i. Fails if y-coordinate of any R is not a quadratic residue or if
+ * 0 != (s1 + a2*s2 + ... + au*su)G - R1 - a2*R2 - ... - au*Ru - e1*P1 - (a2*e2)P2 - ... - (au*eu)Pu. */
+int secp256k1_schnorrsig_verify_batch(const secp256k1_context *ctx, secp256k1_scratch *scratch, const secp256k1_schnorrsig *const *sig, const unsigned char *const *msg32, const secp256k1_pubkey *const *pk, size_t n_sigs) {
+    secp256k1_schnorrsig_verify_ecmult_context ecmult_context;
+    secp256k1_sha256 sha;
+    secp256k1_scalar s;
+    secp256k1_gej rj;
+
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(secp256k1_ecmult_context_is_built(&ctx->ecmult_ctx));
+    ARG_CHECK(scratch != NULL);
+
+    secp256k1_sha256_initialize(&sha);
```

Comment:
Can we pull the `sha` object into `secp256k1_schnorrsig_verify_batch_init_randomizer`?

Reply:
An argument against that is that it makes combining multiple `init_randomizers` less succinct. With a shared sha object it'd look like
```
    secp256k1_sha256_initialize(&sha);
    secp256k1_schnorr_verify_init_randomizer(ctx, &ecmult_data.schnorr, &sha, sig, msg32, pk, n_sigs);
    secp256k1_taproot_verify_init_randomizer(ctx, &sha, ...);
    secp256k1_sha256_finalize(&sha, ecmult_data.schnorr.chacha_seed);
```

Reply:
Gotcha, thanks! I wasn't sure if this was done with an eye toward taproot.